### DISCUSSION TITLE
Generate, translate and format subtitles in one liner

### DISCUSSION THREAD

**Discussion by baxtree** [openai/whisper#1378] [Show and tell]
Hiya, would like to share that whisper has been integrated into [subaligner](https://github.com/baxtree/subaligner)’s subtitle generation process. This integration, along with existing converters and other translation models, enables us to create near-production quality subtitles in a wide range of formats and languages.

Thank you for keeping those fantastic models accessible to everyone.

**Comment by dgoryeo** [comment]
Hi @baxtree , this looks quite interesting. One of the weaknesses of Whisper models is that a subtitle line can start seconds before the actual speech --the timings are not as accurate. Would Subligner be able to address that --correct the start/end times for each subtitle line?  Thanks.

**Reply by ryanheise** [reply]
> One of the weaknesses of Whisper models is that a subtitle line can start seconds before the actual speech

@dgoryeo this should be addressed when `--word_timestamps True` is enabled.

**Reply by baxtree** [reply]
Hi, @dgoryeo, Yes, Subaligner was initially created for subtitle synchronisation. So it is possible to combine generation and synchronisation by chaining multiple commands together, which may help to address timecode issues arising from ASR systems. E.g.,
```
$ subaligner -m transcribe -v video.mp4 -ml eng -mr whisper -mf medium -o may_be_out_of_sync.vtt
$ subaligner -m dual -v video.mp4 -s may_be_out_of_sync.vtt -o subtitle.vtt

**Reply by dgoryeo** [reply]
Thanks @baxtree , is there any dependency to the laguage for dual-stage alignment? My source language is Japanese.

Asid to @ryanheise , although --word_timestamps has improved the time accuracy, there still exists a noticable number of inaccurate timestamped lines --especially for large model. Tuning VAD options helps as well but not completely.

**Reply by ryanheise** [reply]
May I ask, how early (or late) is Whisper starting the line in the worst case case you have observed, where `--word_timestamps` is enabled?

**Reply by baxtree** [reply]
Hi, @dgoryeo, no, both single-stage and dual-stage alignments are language agnostic. However, transcribing and translation are language/language-pair dependent (requiring `subaligner[llm]`) and also true for forced alignment (requiring `subaligner[stretch]`). In your case, the source language code will be `jpn`.

**Reply by dgoryeo** [reply]
Thanks @baxtree , I'll look into it. Much appreciated!

@ryanheise , In my experience the early start of a sub can be anything between 6 - 2 seconds. Very often that happens after a long period of (background) noise but no speech. Funny thing is that the end timestamp is almost always accurate :)

**Reply by ryanheise** [reply]
Do you have a sample file that I can test with in order to develop a fix? With word timestamps enabled, this should not happen so it's a bug worth fixing.

**Reply by dgoryeo** [reply]
@ryanheise , thanks for the suggestion. Let me look for a sample that wouldn't have copyright issues to share. Will come back.

**Reply by dgoryeo** [reply]
Hi @ryanheise , here is a sample file that I hope can be used to diagnose misalignment of the start time (too early start). The audio clip is about 90 seconds from a Japanese anime. I have added the original (production quality) subs for the clip, as well as my Whipser output from the large model. I have packed the audio + the subs in this zip file.

In this clip there was 2 occurances of wide misalignment: one is at about 20sec in to the clip that last for about 10 sec. The other one is at 47sec in to the audio and last about 5 sec.

[sample_zip](https://mega.nz/file/tVcGTC5T#tFo1F3iaGw2hR5I_Lb61hAGSZbYdvb9UeaDTK_eTZR4)
 
My settings for producing the subs are below: 

```
I have used SileroVAD for the entire video [threshhold: 0.25].
Whisper model: large
Whisper parameters:
                        "language":"Japanese", 
                        "verbose":False, 
                        "task":"transcribe", 
                        "word_timestamps":True,
                        "temperature":0, 
                        "best_of":2, 
                        "beam_size":5, 
                        "patience":2, 
                        "length_penalty":0.25, 
                        "suppress_tokens":"-1", 
                        "initial_prompt":promptja, 
                        "condition_on_previous_text":False, 
                        "compression_ratio_threshold":1.9, 
                        "logprob_threshold":-1.0, 
                        "no_speech_threshold":0.15      


```
I hope this helps. Thanks.

**Reply by ryanheise** [reply]
Can you copy and paste the specific line(s) or block numbers from the Whisper SRT you want me to look at?

**Reply by dgoryeo** [reply]
@ryanheise , this line exhibits the problem:

```
17
00:00:36,532 --> 00:00:49,188
```

The correct/exact timing should have been:

```
17
00:00:47,430 --> 00:00:49,266
```

Thanks.

**Reply by ryanheise** [reply]
That's interesting. I might need to wait for the weekend before I can experiment with it, but I am guessing that words in the middle of the segment are elongated rather than just the edge words. Maybe you can confirm this by generating another SRT with word highlighting (i.e. `"highlight_words": True,`).

**Reply by dgoryeo** [reply]
@ryanheise , I think your hunch is right. I ran the writer with the option `writer_options = {"max_line_width": 20, "max_line_count": 2,"highlight_words":True}`. 
Here are the results:

Original line 17 sub:
```

17
00:00:36,532 --> 00:00:49,188
くそ 来やがってな


```
The new lines corresponding to the original line 17 with the writer_options:

```

172
00:00:41,340 --> 00:00:47,160
<u>そ</u>

173
00:00:47,640 --> 00:00:47,800
<u>来</u>やがってな

174
00:00:47,800 --> 00:00:47,920
来<u>や</u>がってな

175
00:00:47,920 --> 00:00:47,980
来や<u>が</u>ってな

176
00:00:47,980 --> 00:00:48,100
来やが<u>って</u>な

177
00:00:48,100 --> 00:00:48,260
来やがって<u>な</u>

```

It looks that whisper (too early) picks up the character "そ" (which seems to be a ghost) and holds on to it for 6 seconds.

**Reply by ryanheise** [reply]
Well that one's on the edge not the middle, which is good news since it should be easier to fix.

**Reply by baixiac** [reply]
Oh cool. Is the potential fix language-specific or not? Currently, we just use the default arg values of `transcribe()` and would be good to know whether the fix is transparent to the client-side call.

**Reply by Soykertje** [reply]
I got better results in this matter by adjusting the `no_speech_threshold` which is an argument of `transcribe()` function, by default it's 0.6 but after playing around with it, just found that set it as 0.4 leads to a better result, specially if the audio file has some background music or some noise... Of course it has its trade-offs but i think that it's worth for many use-cases with that gain into the time-stamps accuracy.

**Reply by dgoryeo** [reply]
Thanks @Soykertje , where might I be able to learn more about the impact / inner workings of `no_speech_threshold`  ? I had set the threshold in my settings to 0.16 which I believe I had blindly picked up from an article in Medium.

**Reply by dgoryeo** [reply]
> Well that one's on the edge not the middle, which is good news since it should be easier to fix.

Hi @ryanheise , by any chance did you get time to take a look at the issue with start-time error? I'm very keen to find some work around it. I wonder if playing with no-speech_threshold as suggested by @Soykertje might help as a workaround? Thanks.

**Reply by ryanheise** [reply]
@dgoryeo Didn't get a chance, unfortunately. If I'm lucky, this coming weekend will give me another chance to look at it.

**Reply by Soykertje** [reply]
@dgoryeo by the `transcribe()` docstring:

> no_speech_threshold: float
        If the no_speech probability is higher than this value AND the average log probability
        over sampled tokens is below `logprob_threshold`, consider the segment as silent

You can find the `transcribe()`'s function implementation [here](https://github.com/openai/whisper/blob/main/whisper/transcribe.py) but as far as I understand, one of the pre-trained model outputs is `no_speech_prob` which is a probability that in that chunk there's not speech.

I forgot to mention in my previous comment but that only option does not increment the accuracy by itself, in my use-cases, setting `word_timestamps=True` in combination with `no_speech_threshold=0.4` is what leads to a better accuracy. 

By the `transcribe()` docstring:

> word_timestamps: bool
        Extract word-level timestamps using the cross-attention pattern and dynamic time warping,
        and include the timestamps for each word in each segment.

But that parameter also affects to other such as `prepend_punctuations` and `append_punctuations`.

IMO at this moment there's no way to improve the accuracy in the timestamps further than that without changing some parts such the sample length (which is 30 seconds), but that can lead to weird behaviors, I hope that in a next release this gets better along with the multi-speaker identification which at this moment does not work very well.

**Reply by ryanheise** [reply]
@dgoryeo finally got a chance to work on this today. I've created a PR #1461 to address this. Let me know how it goes.

**Reply by baxtree** [reply]
The funny thing is after upgrading Whisper from v20230124 to v20230314, I can't see substantial start-time issues for the audio clip shared by @dgoryeo earlier. No additional VAD-based alignments were applied and all arguments of `transcribe()` used the default values except for `language`.

**Reply by dgoryeo** [reply]
Thanks @ryanheise. I noticed that [f572f21](https://github.com/openai/whisper/commit/f572f2161ba831bae131364c3bffdead7af6d210) was merged into openai:main [4 days ago](https://github.com/openai/whisper/pull/1461#event-9682511633).  I will test it later this week to check.

**Reply by baxtree** [reply]
Upgraded to v20230918 and got the same time codes as those with v20230314, better than v20230124. [srt-subtitles.zip](https://github.com/openai/whisper/files/12867250/srt-subtitles.zip)
`pip install openai-whisper==20230918`
`subaligner -m transcribe -v 'ItouJunji-ja-WhisperLarge+SileroVAD-0.25.wav' -ml jpn -mr whisper -mf large -o 'ItouJunji-ja-WhisperLarge-v20230918.srt'`

**Reply by baxtree** [reply]
Update on v20231106 and model large-v3: Some degradation in transcription quality has been observed, at least for `jpn` as discussed around the above sample. Sticking to `-mf large-v2` after upgrading to v20231106 brings better outcomes for this use case.
`subaligner -m transcribe -v 'ItouJunji-ja-WhisperLarge+SileroVAD-0.25.wav' -ml jpn -mr whisper -mf large-v2 -o 'ItouJunji-ja-WhisperLarge-v20231106-large-v2.srt'`
[srt-subtitles.zip](https://github.com/openai/whisper/files/13333701/srt-subtitles.zip)

**Comment by phineas-pta** [comment]
how do you deal with hallucinations from whisper?

**Reply by baxtree** [reply]
Good question. This is indeed a missing jewel atm and we plan to rely on the future releases of Whisper or its variants to do the [right thing](https://github.com/openai/whisper/discussions/679). Until then, QA personnel need to be involved in reviewing subtitles before their distribution. Apart from hallucinations, there are other aspects that need to be addressed to meet the editorial guideline anyway.

**Comment by lslatr23** [comment]
Is it possible to install Subaligner directly in Windows? When running the pip commands in installation guide I receive some errors, so I'am wondering if it is even worth tryin to install under Windows. 

I have Linux but only in VMware as virtual machine, and the problem is that the guest OS (Linux) in VMware cannot access the GPU to accelerate AI workloads.

**Reply by phineas-pta** [reply]
when u have problem, instead of wondering, do:
- google / chatgpt at 1st
- if not work then raise issue in repo with full logs

concerning linux dev in windows, if u dont need gui then dont use virtual machine, use WSL2
